src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs

```diff
@@ -0,0 +1,145 @@
+
+            return document;
+        }
+
+        private static Document PerformDecompilation(Document document, string fullName, Compilation compilation, string assemblyLocation)
+        {
+            // Load the assembly.
+            var pefile = new PEFile(assemblyLocation, PEStreamOptions.PrefetchEntireImage);
+
+            // Initialize a decompiler with default settings.
+            var decompiler = new CSharpDecompiler(pefile, new AssemblyResolver(compilation), new DecompilerSettings());
+            // Escape invalid identifiers to prevent Roslyn from failing to parse the generated code.
+            // (This happens for example, when there is compiler-generated code that is not yet recognized/transformed by the decompiler.)
+            decompiler.AstTransforms.Add(new EscapeInvalidIdentifiers());
+
+            var fullTypeName = new FullTypeName(fullName);
+
+            var decompilerVersion = FileVersionInfo.GetVersionInfo(typeof(CSharpDecompiler).Assembly.Location);
+
+            // Add header to match output of metadata-only view.
+            // (This also makes debugging easier, because you can see which assembly was decompiled inside VS.)
+            var header = $"#region {FeaturesResources.Assembly} {pefile.FullName}" + Environment.NewLine
+                + $"// {assemblyLocation}" + Environment.NewLine
+                + $"// Decompiled with ICSharpCode.Decompiler {decompilerVersion.FileVersion}" + Environment.NewLine
+                + "#endregion" + Environment.NewLine;
+
+            // Try to decompile; if an exception is thrown the caller will handle it
+            var text = decompiler.DecompileTypeAsString(fullTypeName);
+            return document.WithText(SourceText.From(header + text));
+        }
+
+        private async Task<Document> ConvertDocCommentsToRegularComments(Document document, IDocumentationCommentFormattingService docCommentFormattingService, CancellationToken cancellationToken)
+        {
+            var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var newSyntaxRoot = DocCommentConverter.ConvertToRegularComments(syntaxRoot, docCommentFormattingService, cancellationToken);
+
+            return document.WithSyntaxRoot(newSyntaxRoot);
+        }
+
+        private IEnumerable<IFormattingRule> GetFormattingRules(Document document)
+        {
+            return Formatter.GetDefaultFormattingRules(document);
+        }
+
+        private string GetFullReflectionName(INamedTypeSymbol containingType)
+        {
+            var stack = new Stack<string>();
+            stack.Push(containingType.MetadataName);
+            var ns = containingType.ContainingNamespace;
+            do
+            {
+                stack.Push(ns.Name);
+                ns = ns.ContainingNamespace;
+            }
+            while (ns != null && !ns.IsGlobalNamespace);
+
+            return string.Join(".", stack);
```

Comment:
nit.  my preference would be that we just use a SymbolDisplayFormat here.  But i may be wrong given that you're using the .MetadataName.  Is that necessary?  can you document why?

Reply:
The method is called `GetFullReflectionName`.... I wonder, isn't that enough documentation? As for the reason: Well... the decompiler API expects the reflection name. :)